Massive clean up for conditional operations

glwagner opened the pull request:
This PR greatly simplifies the logic behind `conditional_operation` and how its extended for immersd boundary grid. The end result besides source code clean up seems to be an improvement in type inference.

Resolves #3750 I think. But @ali-ramadhan please test.

EDIT: I decided not to work on #3791 (here) because this requires fixing a bunch of tests and is a bigger effort.

glwagner commented:
AUUUGHHH that was hard but finally found the source of the type instability:

https://github.com/CliMA/Oceananigans.jl/blob/4dbec269c7fa88a611121a78924589945b535d2d/src/AbstractOperations/AbstractOperations.jl#L37-L44

(`axes` is called with `mapreducedim!` at some point)

fixing that finally made the difference. Just required generalizing the one we already have for `field` (supposed to fix the same problem, except only for `Field` and not abstract operations):

https://github.com/CliMA/Oceananigans.jl/blob/4dbec269c7fa88a611121a78924589945b535d2d/src/Fields/field.jl#L401-L413

navidcy commented:
> AUUUGHHH that was hard but finally found the source of the type instability:

which type instability; is this related to discussion in https://github.com/CliMA/Oceananigans.jl/issues/3750?

glwagner commented:
> > AUUUGHHH that was hard but finally found the source of the type instability:
> 
> which type instability; is this related to discussion in #3750?

Yes. That issue documents slow reductions for windowed fields on immersed boundary grids. I hypothesized that it was due to a failure of type inference. Looking into it further I see that `axes(op::AbstractOperation)` cannot be type inferred when indices is not `(:, :, :)` because of the tuple generator. We found the same problem with `axes` for `Field` and fixed it but didn't fix it for AbstractOperations. Reducing windowed fields on immersed boundary grids requires this because they are wrapped in `ConditionalOperation` in order to mask the immersed regions during the reduction. This PR extends the fix we implemented for `Field` to also encompass `AbstractOperation`. It also cleans up conditional operations quite a bit.

glwagner commented:
It's not really a bug btw

glwagner commented:
But definitely clean up and performance. Just to clarify fixing type inference doesn't change the result of the reduction, it just makes it go much faster

ali-ramadhan commented:
Was OOO last week but can review this later today!

This PR does resolve the slowness in #3750!

```julia
julia> @time minimum(u2)
  0.059865 seconds (665 allocations: 54.930 KiB)
0.0

julia> @time minimum(u2.data)
  0.011223 seconds (3 allocations: 1.562 KiB)
0.0
```

Some overhead which is expected I think, but _way_ better than 2000x slower lol.

glwagner commented:
@ali-ramadhan let's focus first on #3801 and then we can revisit this PR. There are some additional challenges to resolving #3791 unfortunately that will also require some test refactoring.

simone-silvestri commented on src/AbstractOperations/conditional_operations.jl:
```diff
@@ -95,29 +98,44 @@ function ConditionalOperation(c::ConditionalOperation;
 end
 
-struct TrueCondition end
+@inline function Base.getindex(co::ConditionalOperation, i, j, k)
```
```suggestion
@propagate_inbounds function Base.getindex(co::ConditionalOperation, i, j, k)
```
I think

simone-silvestri commented on src/AbstractOperations/conditional_operations.jl:
```diff
@@ -95,29 +98,44 @@ function ConditionalOperation(c::ConditionalOperation;
+# Conditions: general, nothing, array
+@inline evaluate_condition(condition, i, j, k, grid, args...) = condition(i, j, k, grid, args...)
+@inline evaluate_condition(::Nothing, i, j, k, grid, args...) = true
+@propagate_inbounds evaluate_condition(condition::AbstractArray, i, j, k, grid, args...) = condition[i, j, k]
```
```suggestion
@inline evaluate_condition(condition::AbstractArray, i, j, k, grid, args...) = @inbounds condition[i, j, k]
```
or is it the same?

glwagner replied:
It's not exactly the same --- with @propagate_inbounds we do still have to write `@inbouds` somewhere. But we write it at a higher level

simone-silvestri reviewed (approved):
This should be ready to merge?

navidcy commented on src/Fields/abstract_field.jl:
```diff
@@ -126,3 +124,22 @@ for f in (:+, :-)
+
+const XYZReducedAF = AbstractField{Nothing, Nothing, Nothing}
+
+reduced_dimensions(field::AbstractField) = ()
+reduced_dimensions(field::XReducedAF)    = tuple(1)
+reduced_dimensions(field::YReducedAF)    = tuple(2)
+reduced_dimensions(field::ZReducedAF)    = tuple(3)
+reduced_dimensions(field::YZReducedAF)   = (2, 3)
+reduced_dimensions(field::XZReducedAF)   = (1, 3)
+reduced_dimensions(field::XYReducedAF)   = (1, 2)
+reduced_dimensions(field::XYZReducedAF)  = (1, 2, 3)
```
then we should remove these from https://github.com/CliMA/Oceananigans.jl/blob/79e0867b2705eae6691d7a05fe17ad33d4341c5c/src/Fields/field.jl#L521-L528

glwagner replied:
for sure!

glwagner replied:
good catch

glwagner commented on src/AbstractOperations/conditional_operations.jl:
```diff
@@ -109,29 +113,44 @@ function ConditionalOperation(c::ConditionalOperation;
+    conditioned = evaluate_condition(co.condition, i, j, k, c.grid, c)
+    value = getindex(co.operand, i, j, k)
+    func_value = co.func(value)
+    return ifelse(conditioned, value, c.mask)
```
```suggestion
    return ifelse(conditioned, func_value, c.mask)
```

navidcy commented on src/Fields/field.jl:
```diff
@@ -706,10 +699,10 @@ for reduction in (:sum, :maximum, :minimum, :all, :any, :prod)
             loc = reduced_location(location(c); dims)
             r = Field(loc, c.grid, T; indices=indices(c))
-            conditioned_c = condition_operand(f, c, condition, mask)
             initialize_reduced_field!(Base.$(reduction!), identity, r, conditioned_c)
             Base.$(reduction!)(identity, r, conditioned_c, init=false)
```
@glwagner, shouldn't this be:

```suggestion
            initialize_reduced_field!(Base.$(reduction!), f, r, conditioned_c)
            Base.$(reduction!)(f, r, conditioned_c, init=false)
```

navidcy replied:
Does it not need `f` there because `f` is incorporated in `conditioned_c`?

navidcy commented:
```Julia
julia> using Oceananigans
julia> grid = RectilinearGrid(size = (2, 2, 2), x = (0, 2), y = (0, 2), z = (0, 2), topology = (Periodic, Periodic, Bounded))
2×2×2 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 2×2×2 halo
├── Periodic x ∈ [0.0, 2.0) regularly spaced with Δx=1.0
├── Periodic y ∈ [0.0, 2.0) regularly spaced with Δy=1.0
└── Bounded  z ∈ [0.0, 2.0] regularly spaced with Δz=1.0

julia> T = CenterField(grid)
2×2×2 Field{Center, Center, Center} on RectilinearGrid on CPU
├── grid: 2×2×2 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 2×2×2 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: ZeroFlux, top: ZeroFlux, immersed: ZeroFlux
└── data: 6×6×6 OffsetArray(::Array{Float64, 3}, -1:4, -1:4, -1:4) with eltype Float64 with indices -1:4×-1:4×-1:4
    └── max=0.0, min=0.0, mean=0.0

julia> set!(T, (x, y, z) -> x + y + z)
2×2×2 Field{Center, Center, Center} on RectilinearGrid on CPU
├── grid: 2×2×2 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 2×2×2 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: ZeroFlux, top: ZeroFlux, immersed: ZeroFlux
└── data: 6×6×6 OffsetArray(::Array{Float64, 3}, -1:4, -1:4, -1:4) with eltype Float64 with indices -1:4×-1:4×-1:4
    └── max=4.5, min=1.5, mean=3.0

julia> @compute Txyz = Field(Average(T))
1×1×1 Field{Nothing, Nothing, Nothing} reduced over dims = (1, 2, 3) on RectilinearGrid on CPU
├── data: OffsetArray{Float64, 3, Array{Float64, 3}}, size: (1, 1, 1)
├── grid: 2×2×2 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 2×2×2 halo
├── operand: Average of 2×2×2 Field{Center, Center, Center} on RectilinearGrid on CPU over dims (1, 2, 3)
├── status: time=0.0
└── data: 1×1×1 OffsetArray(::Array{Float64, 3}, 1:1, 1:1, 1:1) with eltype Float64 with indices 1:1×1:1×1:1
    └── max=3.0, min=3.0, mean=3.0

julia> @compute Txyz = Field(Average(T, condition=T.>3))
1×1×1 Field{Nothing, Nothing, Nothing} reduced over dims = (1, 2, 3) on RectilinearGrid on CPU
├── data: OffsetArray{Float64, 3, Array{Float64, 3}}, size: (1, 1, 1)
├── grid: 2×2×2 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 2×2×2 halo
├── operand: Average of ConditionalOperation of 2×2×2 Field{Center, Center, Center} on RectilinearGrid on CPU with condition 2×2×2 Array{Bool, 3} over dims (1, 2, 3)
├── status: time=0.0
└── data: 1×1×1 OffsetArray(::Array{Float64, 3}, 1:1, 1:1, 1:1) with eltype Float64 with indices 1:1×1:1×1:1
    └── max=3.0, min=3.0, mean=3.0
```

navidcy commented:
Now it seems like we broke the conditions? Look at this? Adding the condition seems to have no effect.

```Julia
julia> using Oceananigans

julia> grid = RectilinearGrid(arch, size = (2, 2, 2), x = (0, 2), y = (0, 2), z = (0, 2), topology = (Periodic, Periodic, Bounded))
2×2×2 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 2×2×2 halo
├── Periodic x ∈ [0.0, 2.0) regularly spaced with Δx=1.0
├── Periodic y ∈ [0.0, 2.0) regularly spaced with Δy=1.0
└── Bounded  z ∈ [0.0, 2.0] regularly spaced with Δz=1.0

julia> T = CenterField(grid)
2×2×2 Field{Center, Center, Center} on RectilinearGrid on CPU
├── grid: 2×2×2 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 2×2×2 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: ZeroFlux, top: ZeroFlux, immersed: ZeroFlux
└── data: 6×6×6 OffsetArray(::Array{Float64, 3}, -1:4, -1:4, -1:4) with eltype Float64 with indices -1:4×-1:4×-1:4
    └── max=0.0, min=0.0, mean=0.0

julia> set!(T, f_array)^C

julia> set!(T, (x, y, z) -> x + y + z)
2×2×2 Field{Center, Center, Center} on RectilinearGrid on CPU
├── grid: 2×2×2 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 2×2×2 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: ZeroFlux, top: ZeroFlux, immersed: ZeroFlux
└── data: 6×6×6 OffsetArray(::Array{Float64, 3}, -1:4, -1:4, -1:4) with eltype Float64 with indices -1:4×-1:4×-1:4
    └── max=4.5, min=1.5, mean=3.0

julia> @compute Txyz = Field(Average(T))
1×1×1 Field{Nothing, Nothing, Nothing} reduced over dims = (1, 2, 3) on RectilinearGrid on CPU
├── data: OffsetArray{Float64, 3, Array{Float64, 3}}, size: (1, 1, 1)
├── grid: 2×2×2 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 2×2×2 halo
├── operand: Average of 2×2×2 Field{Center, Center, Center} on RectilinearGrid on CPU over dims (1, 2, 3)
├── status: time=0.0
└── data: 1×1×1 OffsetArray(::Array{Float64, 3}, 1:1, 1:1, 1:1) with eltype Float64 with indices 1:1×1:1×1:1
    └── max=3.0, min=3.0, mean=3.0

julia> @compute Txyz = Field(Average(T, condition=T.>3))
1×1×1 Field{Nothing, Nothing, Nothing} reduced over dims = (1, 2, 3) on RectilinearGrid on CPU
├── data: OffsetArray{Float64, 3, Array{Float64, 3}}, size: (1, 1, 1)
├── grid: 2×2×2 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 2×2×2 halo
├── operand: Average of ConditionalOperation of 2×2×2 Field{Center, Center, Center} on RectilinearGrid on CPU with condition 2×2×2 Array{Bool, 3} over dims (1, 2, 3)
├── status: time=0.0
└── data: 1×1×1 OffsetArray(::Array{Float64, 3}, 1:1, 1:1, 1:1) with eltype Float64 with indices 1:1×1:1×1:1
    └── max=3.0, min=3.0, mean=3.0
```

navidcy commented:
As soon as there are non-trivial indices the

```
julia> mean(interior(c, :, :, :)) ≈ mean(view(c, :, :, :))
true

julia> mean(interior(c, :, :, 1:1)) ≈ mean(view(c, :, :, 1:1))
false
```

```
julia> using Oceananigans.AbstractOperations: condition_operand, conditional_length

julia> operator = condition_operand(identity, view(c, :, :, :), nothing, 0); conditional_length(operator)
24

julia> operator = condition_operand(identity, view(c, :, :, 1:1), nothing, 0); conditional_length(operator)
24
```

navidcy commented:
I think the problem is with the `conditional_one`:

https://github.com/CliMA/Oceananigans.jl/blob/7c429d3a68474abacc9b6b52f703194642bf4b89/src/AbstractOperations/conditional_operations.jl#L179-L184

But I couldn't figure out a fix.

navidcy commented:
This is I believe one of the times that the doctests picked up something that the tests missed!

glwagner commented:
> As soon as there are non-trivial indices the
> 
> ```julia
> julia> mean(interior(c, :, :, :)) ≈ mean(view(c, :, :, :))
> true
> 
> julia> mean(interior(c, :, :, 1:1)) ≈ mean(view(c, :, :, 1:1))
> false
> ```
> 
> ```julia
> julia> using Oceananigans.AbstractOperations: condition_operand, conditional_length
> 
> julia> operator = condition_operand(identity, view(c, :, :, :), nothing, 0); conditional_length(operator)
> 24
> 
> julia> operator = condition_operand(identity, view(c, :, :, 1:1), nothing, 0); conditional_length(operator)
> 24
> ```
> 
> despite that the
> 
> ```julia
> julia> length(view(c, :, :, 1:1))
> 6
> ```

Is the last result right? there are only 6 points?

glwagner commented:
> yes, the result should be 6 but conditional_length gives 24; seems like conditional_length still "sees" the original field `c` with all indices (Colon(), Colon(), Colon())

ok, so we need to fix `conditional_length` then right?

glwagner commented:
also, an additional test seems warranted

glwagner commented on src/AbstractOperations/conditional_operations.jl:
```diff
@@ -176,12 +176,26 @@ function materialize_condition(c::ConditionalOperation)
+    LX, LY, LZ = location(c)
+    grid = c.operand.grid
+    indices = c.operand.indices
+    one_field = Field{LX, LY, LZ}(grid; indices)
+    set!(one_field, 1)
```
```suggestion
    if indices == (:, :, :)
        one_field = OneField(Int)
    else
        one_field = Field{LX, LY, LZ}(grid; indices)
        set!(one_field, 1)
    end
```

navidcy replied:
@glwagner I essentially already did what you suggest but via dispatch; see: https://github.com/CliMA/Oceananigans.jl/blob/50131275c1e7ae344abcdfd3933bab423f677a09/src/AbstractOperations/conditional_operations.jl#L179-L197

Did you miss that? Or there is a way to prefer using `if else`?

glwagner replied:
provided that type inference succeeds, the if-statement is identical to manual dispatch in terms of generated code

glwagner replied:
so its up to you which you find to be easier to read and maintain

navidcy replied:
(from Slack chat)

glwagner: the two are identical because the type of `indices` is known. so its up to you in terms of readability and maintainability, i think the if statement is a little clearer to read

navidcy replied:
agree, if is cleaner!

navidcy commented:
@glwagner I essentially already did what you suggest but via dispatch; see: https://github.com/CliMA/Oceananigans.jl/blob/50131275c1e7ae344abcdfd3933bab423f677a09/src/AbstractOperations/conditional_operations.jl#L179-L197

Did you miss that? Or there is a way to prefer using `if else`?

glwagner commented on test/test_field.jl:
```diff
@@ -471,6 +471,21 @@ end
+            c = CenterField(grid)
+
+            Random.seed!(42)
+            c .= rand(size(c)...)
```
```suggestion
            set!(c,  rand(size(c)...))

```

glwagner commented on test/test_field.jl:
```diff
@@ -471,6 +471,21 @@ end
         for arch in archs, FT in float_types
             run_field_reduction_tests(FT, arch)
+
+            # test reductions on windowed fields
+
+            grid = RectilinearGrid(arch, FT, size=(2, 3, 4), x=(0, 1), y=(0, 1), z=(0, 1))
+
+            c = CenterField(grid)
+
+            Random.seed!(42)
```
```suggestion
            @info "    Test reductions on WindowedFields...
            grid = RectilinearGrid(arch, FT, size=(2, 3, 4), x=(0, 1), y=(0, 1), z=(0, 1))
            c = CenterField(grid)
            Random.seed!(42)
```

glwagner commented on test/test_field.jl:
```diff
@@ -471,6 +471,19 @@ end
         for arch in archs, FT in float_types
             run_field_reduction_tests(FT, arch)
+
+            @info "    Test reductions on WindowedFields...
```
```suggestion
            @info "    Test reductions on WindowedFields..."

```

glwagner commented on test/test_field.jl:
```diff
@@ -471,6 +471,20 @@ end
+            grid = RectilinearGrid(arch, FT, size=(2, 3, 4), x=(0, 1), y=(0, 1), z=(0, 1))
+            c = CenterField(grid)
+            Random.seed!(42)
+            set!(c,  rand(size(c)...))
```
```suggestion
            set!(c, rand(size(c)...))

```

glwagner commented on test/test_field.jl:
```diff
@@ -471,6 +471,21 @@ end
+            c = CenterField(grid)
+            Random.seed!(42)
+            set!(c, rand(size(c)...))
+
```
```suggestion
```

glwagner commented on test/test_field.jl:
```diff
@@ -471,6 +471,20 @@ end
+            grid = RectilinearGrid(arch, FT, size=(2, 3, 4), x=(0, 1), y=(0, 1), z=(0, 1))
+            c = CenterField(grid)
+            Random.seed!(42)
+            set!(c, rand(size(c)...))
```
PS @navidcy this line used to use broadcasting, eg ` c .= ...`. Note that broadcasting with Fields has limited support, so probably not good for testing (or anywhere really... would be nice to make it better but there's lots to do...)

navidcy replied:
I think I took it from the doctest. I’ll change it there!
